File: Java/math/Matrix_multiplication.java
/* MATRIX MULTIPLICATION
    2 matrices are input and multiplied using following algorithm:
    START
    Input no. of rows m1 and columns n1 of first matrix
    For i = 0 to m1-1
        For j=0 to n1-1
            Input elements of first matrix and store in a[i][j]
    Input no. of rows m2 and columns n2 of second matrix
    For i = 0 to m2-1
        For j=0 to n2-1
            Input elements of second matrix and store in b[i][j]
    If (n1 != m2)
        Print "Multiplication is not possible"
    Else
        Initialize product matrix c[m1][n2] as 0 matrix
        For i = 0 to m1-1
            For j=0 to n2-1
                For k=0 to n1-1
                    c[i][j] =c[i][j]+( a[i][k] * b[k][j])
        For i = 0 to m1-1
            For j=0 to n2-1
                Print c[i][j] 
    STOP 

*/

//import package to accept input
import java.util.Scanner;

//class for matrix multiplication
public class Matrix {

    ////to multiply 2 matrices.2 matrices and corresponding rows and columns are arguments
    static void multiply(int a[][],int b[][],int m1,int n1,int m2,int n2){
        //check if no of columns of first matrix is 
        //equal to number of rows of second matrix 
        if (n1 != m2)
            System.out.println("Multiplication is not possible");

        else 
        {
            int i,j;
            //initialize new matrix to store product
            int c[][] = new int[m1][n2];

            //matrix multiplication
            for (i = 0; i < m1; i++) 
            {
                for (j = 0; j < n2; j++) 
                {
                    for (int k = 0; k < n1; k++)
                        c[i][j] += a[i][k] * b[k][j];
                }
            }
            //to print product matrix 
            System.out.println("Product:");
            for (i = 0; i < m1; i++) 
            {
                for (j = 0; j < n2; j++) 
                {
                    System.out.print(c[i][j]+" " );
                }
                System.out.println();
            }
        }
    }

    //driver code
    public static void main(String[] args) {
        int i, j, temp;
        //accept the matrices, rows and columns as user input
        Scanner s = new Scanner(System.in);
        System.out.print("Enter the number of rows and columns in first matrix:");
        int m1 = s.nextInt();
        int n1 = s.nextInt();
        int a[][] = new int[m1][n1];
        System.out.print("Enter the elements of first matrix:");
        for (i = 0; i < m1; i++) 
        {
            for (j = 0; j < n1; j++)
                a[i][j] = s.nextInt();
        }
        System.out.print("Enter the number of rows and columns in second matrix:");
        int m2 = s.nextInt();
        int n2 = s.nextInt();
        int b[][] = new int[m2][n2];
        System.out.print("Enter the elements of second matrix:");
        for (i = 0; i < m2; i++) 
        {
            for (j = 0; j < n2; j++)
                b[i][j] = s.nextInt();
        }
        //call function to multiply the matrices
        multiply(a,b,m1,n1,m2,n2);
    }
}

/*  Sample I/O:
    Enter the number of rows and columns in first matrix:2 3                                                                                      
    Enter the elements of first matrix:1 2 3 4 5 6                                                                                                
    Enter the number of rows and columns in second matrix:3 2                                                                                     
    Enter the elements of second matrix:1 2 3 4 5 6                                                                                               
    Product:                                                                                                                                      
    22 28                                                                                                                                         
    49 64  
*/
/* Time Complexity: O(m1*n2*n1)
   Space Complexity: O(m1*n2)
*/

File: C/math/matrix_multiplication.c
// C program to implement Matrix Multiplication
#include <stdio.h>
void matrix_mult(int [][10],int [][10],int [][10],int,int,int,int);
int main()
{
    int r1, c1, r2, c2;
    printf("Enter the number of rows and columns of the first matrix:  ");
    scanf("%d %d", &r1, &c1);
    printf("Enter the number of rows and columns of the second matrix: ");
    scanf("%d %d", &r2, &c2);

    //If the number of columns of the second matrix and the number of rows of first matrix differ, they cannot be added
    if (c1 != r2)
    {
        printf("Given Matrices cannot be multiplyable!!!");
        return 0;
    }

    int A[10][10], B[10][10], C[10][10];
    // Input the values of the matrices
    printf("Enter the values of the first matrix\n");
    for (int i = 0; i < r1; i++)
    {
        for (int j = 0; j < c1; j++)
            scanf("%d", &A[i][j]);
    }
    printf("Enter the values of the second matrix\n");
    for (int i = 0; i < r2; i++)
    {
        for (int j = 0; j < c2; j++)
            scanf("%d", &B[i][j]);
    }
    matrix_mult(C,A,B,r1,r2,c1,c2);
    printf("The resultant matrix is:\n");
    for (int i = 0; i < r2; i++)
    {
        for (int j = 0; j < c2; j++)
            printf("%d ", C[i][j]);
        printf("\n");
    }
    return 0;
}
void matrix_mult(int C[][10],int A[][10],int B[][10],int r1,int r2,int c1,int c2)
{
    // Multiply both the matrices
    for (int i = 0; i < r1; i++)
    {
        for (int j = 0; j < c2; j++)
        {
            C[i][j] = 0;
            for (int k = 0; k < c1; k++)
                C[i][j] += A[i][k] * B[k][j];
        }
    }

}

/*
Time Complexity: O(r1 * c2 * c1), where 'r1' is the number of rows of first matrix and 'c2' is the number of columns 
                 of second matrix and 'c1' is the number of columns of first matrix   
Space Complexity: O(r1 * c2)

SAMPLE INPUT AND OUTPUT

Enter the number of rows and columns of the first matrix:  2 2
Enter the number of rows and columns of the second matrix: 2 2
Enter the values of the first matrix
1 1
1 1
Enter the values of the second matrix
1 1
1 1
The resultant matrix is:
2 2
2 2

*/

